hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/oauth2/WorkloadIdentityTokenProvider.java

```diff
@@ -0,0 +1,144 @@
+import org.apache.hadoop.util.Preconditions;
+
+/**
+ * Provides tokens based on Azure AD Workload Identity.
+ */
+public class WorkloadIdentityTokenProvider extends AccessTokenProvider {
+
+  private static final String OAUTH2_TOKEN_PATH = "/oauth2/v2.0/token";
+  private static final long ONE_HOUR = 3600 * 1000;
+  private static final Logger LOG = LoggerFactory.getLogger(AccessTokenProvider.class);
+
+  private final String authEndpoint;
+  private final String clientId;
+  private final String tokenFile;
+  private long tokenFetchTime = -1;
+
+  public WorkloadIdentityTokenProvider(final String authority, final String tenantId,
+      final String clientId, final String tokenFile) {
+    Preconditions.checkNotNull(authority, "authority");
+    Preconditions.checkNotNull(tenantId, "tenantId");
+    Preconditions.checkNotNull(clientId, "clientId");
+    Preconditions.checkNotNull(tokenFile, "tokenFile");
+
+    this.authEndpoint = authority + tenantId + OAUTH2_TOKEN_PATH;
+    this.clientId = clientId;
+    this.tokenFile = tokenFile;
+  }
+
+  @Override
+  protected AzureADToken refreshToken() throws IOException {
+    LOG.debug("AADToken: refreshing token from JWT Assertion");
+    String clientAssertion = getClientAssertion();
+    AzureADToken token = getTokenUsingJWTAssertion(clientAssertion);
+    tokenFetchTime = System.currentTimeMillis();
+    return token;
+  }
+
+  /**
+   * Gets the Azure AD token from a client assertion in JWT format.
+   * This method exists to make unit testing possible.
+   *
+   * @param clientAssertion the client assertion.
+   * @return the Azure AD token.
+   * @throws IOException if there is a failure in connecting to Azure AD.
+   */
+  @VisibleForTesting
+  AzureADToken getTokenUsingJWTAssertion(String clientAssertion) throws IOException {
+    return AzureADAuthenticator
+        .getTokenUsingJWTAssertion(authEndpoint, clientId, clientAssertion);
+  }
+
+  /**
+   * Checks if the token is about to expire as per base expiry logic.
+   * Otherwise, try to expire if enough time has elapsed since the last refresh.
+   *
+   * @return true if the token is expiring in next 1 hour or if a token has
+   * never been fetched
+   */
+  @Override
+  protected boolean isTokenAboutToExpire() {
+    return super.isTokenAboutToExpire() || hasEnoughTimeElapsedSinceLastRefresh();
+  }
+
+  /**
+   * Checks to see if enough time has elapsed since the last token refresh.
+   *
+   * @return true if the token was last refreshed more than an hour ago.
+   */
+  protected boolean hasEnoughTimeElapsedSinceLastRefresh() {
+    if (getTokenFetchTime() == -1) {
+      return true;
+    }
+    boolean expiring = false;
+    long elapsedTimeSinceLastTokenRefreshInMillis =
+        System.currentTimeMillis() - getTokenFetchTime();
+    // In case token is not refreshed for 1 hr or any clock skew issues,
+    // refresh token.
+    expiring = elapsedTimeSinceLastTokenRefreshInMillis >= ONE_HOUR
```

Comment:
This is the same logic used today for MSI Tokens as well. I guess original author ported the logic from there itself.
Also, as a recent change made by Azure services, the lifetime of an azure token was made variable, earlier I guess it was 1 hour only, which lead to hourly spikes on servers for refresh requests.

It seems better to rely only on the "expires_on" parameter (already done in base class) returned along with token by servers. But this will need proper backing before changing as MSI Token code has been there for a while.

If it seems okay, I would prefer relooking at this in a separate Work Item (Jira ticket)
Let me know your thoughts on this.

Reply:
Removed the 1 hour force expiry and refresh of token. Will rely only on the expiration time set by the server on the token returned.